File: Lab1/lab1_algo/src/main/java/com/example/Reader.java
package com.example;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class Reader {

    public static String md5Hash(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] hash = md.digest(input.getBytes());

            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                hexString.append(String.format("%02x", b));
            }
            return hexString.toString();

        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("MD5 not supported", e);
        }
    }

    public static String readFile(String fileName) {
        StringBuilder content = new StringBuilder();

        try (BufferedReader reader = new BufferedReader(new FileReader(fileName))) {
            String line;
            boolean isEmpty = true;

            while ((line = reader.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    isEmpty = false;
                }
                content.append(line).append("\n");
            }

            if (isEmpty) {
                throw new RuntimeException("File is empty!");
            }

            return content.toString();

        } catch (IOException e) {
            throw new RuntimeException("Error to read file: " + e.getMessage(), e);
        }
    }
}

File: Lab1/Cpp/src/Reader.cpp
#include "Reader.h"

std::string Reader::readFile(const std::string &fileName)
{
    std::ifstream file(fileName);
    if (!file.is_open())
    {
        throw std::runtime_error("Error to read file: " + fileName);
    }

    std::string content;
    std::string line;
    bool empty = true;

    while (std::getline(file, line))
    {
        if (!line.empty() && line.find_first_not_of(" \t\n\r") != std::string::npos)
        {
            empty = false;
        }
        content += line + "\n";
    }

    if (empty)
    {
        return "file is empty";
    }

    return content;
}